src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
> I was thinking of tests that would attempt to access an ontology from multiple....

Ah right, more than a unit test then.  Sorry, I don't have these.

> Regarding listeners and other ancillary collections in the managers, I'd be more tempted to synchronize on the collection itself, 

Well, what ever is equivalent.

> or use synchronized collections 

I'm not sure this is a good idea.

> Also, they are completely separate from ontology data, I believe - unless we want to guarantee that a listener receives all changes sent by a thread up until it gets removed from the manager by another thread?

Yes, something like this. 

> My head is spinning a bit now, I think we're describing two different sets of changes.

:)